Add evictionPolicy field for spot VMs

willie-yao opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature

<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
-->

**What this PR does / why we need it**:
This PR adds a field `evictionPolicy` to `spotVMOptions`. This can either be set to Delete or Deallocate.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2570

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add evictionPolicy field for spot VMs
```

CecileRobertMichon commented on api/v1beta1/azuremachine_types.go:
```diff
@@ -133,6 +134,10 @@ type SpotVMOptions struct {
+
+	// EvictionPolicy defines the behavior of the virtual machine when it is evicted. It can be either Delete or Deallocate.
+	// +optional
+	EvictionPolicy *compute.VirtualMachineEvictionPolicyTypes `json:"evictionPolicy,omitempty"`
```
we want to avoid adding a dependency on the Azure SDK in the api package. This is so that the SDK can we swapped out easily for another (for example this might be applicable to support other clouds besides Azure public cloud). Here instead of using `compute.VirtualMachineEvictionPolicyTypes` you can define a new type

CecileRobertMichon commented:
/ok-to-test

CecileRobertMichon reviewed (commented):
/lgtm

jackfrancis commented:
/test ls

jackfrancis commented:
/test pull-cluster-api-provider-azure-e2e-workload-upgrade

jackfrancis commented:
/retest

willie-yao commented:
/retest

jackfrancis commented:
FYI we actually have to update the service code flow so this new feature actually does something. Search for `SpotVMOptions` under `azure/scope/*` and `/azure/services/*` to see how this is done for the existing spot priority config for vm/vmss.

Also we want to update the docs in `docs/book/src/topics/spot-vms.md`.

willie-yao commented:
I tested this option through custom templates on Tilt and it was able to be configured. Does the feature not work for you?

willie-yao commented:
/retest

willie-yao commented:
/retest

CecileRobertMichon commented on docs/book/src/topics/spot-vms.md:
````diff
@@ -48,6 +48,18 @@ spec:
 ```
 
+In addition, you are able to explicitly set the eviction policy for the Spot VM.
+The default policy is `Delete` which will delete the VM when it is evicted. You
````
should that say "The default policy is `Deallocate`"? https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2689/files#diff-2128b31583d63daf5c4f86f34b7f90859ec634490812674658d179694a88bf2bR44

CecileRobertMichon commented on azure/converters/spotinstances.go:
```diff
@@ -40,8 +40,12 @@ func GetSpotVMOptions(spotVMOptions *infrav1.SpotVMOptions, diffDiskSettings *in
 	}
+
+	// Set the spot vm eviction policy if provided, otherwise default to deallocate for regular disks and delete for ephemeral disks
 	evictionPolicy := compute.VirtualMachineEvictionPolicyTypesDeallocate
```
rethinking this, it would be better to explicitly default the field in the defaulting webhook so that the value is persisted in the Spec, even if the user didn't explicitly set it. That way, if we decide to change the default later, we won't break existing users.

CecileRobertMichon replied:
Webhook defaulting logic for AzureMachines: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremachine_default.go
Webhook defaulting logic for AzureMachinePools: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/exp/api/v1beta1/azuremachinepool_default.go

More info on webhooks: https://cluster-api.sigs.k8s.io/developer/providers/webhooks.html

willie-yao replied:
If we define the default in the webhook, does that mean we won't have to do any checks inside the spotinstances converter, and just return the eviction policy from the spotVMOptions parameter directly? I'm assuming that by the time the converter is called when building the VM from spec, the webhook will set the default spotVMOptions if it isn't defined in the spec.

willie-yao replied:
I'm also wondering if we should still default to delete if an ephemeral disk is being used. If the spec is already defined by the webhook by default, I'm not sure how to check if that's the case before setting it to delete for an ephemeral disk.

CecileRobertMichon replied:
> If we define the default in the webhook, does that mean we won't have to do any checks inside the spotinstances converter, and just return the eviction policy from the spotVMOptions parameter directly

correct

> I'm also wondering if we should still default to delete if an ephemeral disk is being used

you can make the webhook default a "smart" default where it defaults to delete if ephemeral disk is being used

willie-yao replied:
@CecileRobertMichon I was able to make changes to the webhooks to set the default spot eviction policy. However, many of the unit tests started failing because I don't think the defaulting webhook is run for them. Is there a way to run the webhook in a unit test environment?

CecileRobertMichon replied:
see my comment below https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2689/files#r990558486

CecileRobertMichon commented:
> FYI we actually have to update the service code flow so this new feature actually does something

I had the same initial reaction when reviewing this, but in this case it's actually not necessary because the whole SpotVMOptions struct is being passed through to the service (whether we should be doing this is a separate discussion...) so it's already getting the new spot VM option without any additional changes needed to the spec: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/scope/machine.go#L165

CecileRobertMichon commented on azure/converters/spotinstances.go:
```diff
@@ -40,9 +40,9 @@ func GetSpotVMOptions(spotVMOptions *infrav1.SpotVMOptions, diffDiskSettings *in
-	}
+
+	// Set the spot vm eviction policy if provided.
+	evictionPolicy := compute.VirtualMachineEvictionPolicyTypes(*spotVMOptions.EvictionPolicy)
```
we want to make sure we guarding against any potential nil pointer dereferencing here (even though we don't expect this to ever be nil in practice)

```
// Set the spot vm eviction policy if provided.
if spotVMOptions != nil {
    evictionPolicy := compute.VirtualMachineEvictionPolicyTypes(*spotVMOptions.EvictionPolicy)
}
```


This will also fix the unit test failures you're seeing along the lines of `panic: runtime error: invalid memory address or nil pointer dereference`

willie-yao replied:
This makes sense! I implemented this, but the unit tests that test if the spot vm converter returns a default eviction policy is now broken, since the defaults are being set in the webhook. Should I just remove the [test](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2a0afb510a3554f4aacfaa411d9c66cd2245ecbe/azure/services/virtualmachines/spec_test.go#L273) entirely?

willie-yao commented:
Just rebased on main. Not sure why but Git is saying I changed [azure_vmextensions.go](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2689/files#diff-2bea7aae68fda3b77b9d0180bd026269d0804ded551d4d9d5897e528cf20da31) with the same changes from #2698 . It must be a bug since main has the changes shown on the diff for this PR.

CecileRobertMichon reviewed (commented):
/lgtm
/assign @jackfrancis

CecileRobertMichon commented:
> Not sure why but Git is saying I changed [azure_vmextensions.go](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2689/files#diff-2bea7aae68fda3b77b9d0180bd026269d0804ded551d4d9d5897e528cf20da31) with the same changes from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2698 . It must be a bug since main has the changes shown on the diff for this PR.

hmm that's strange... could you please try git fetch and git rebase again?

jackfrancis reviewed (approved):
/approve
